ui/lib/core/addon/helpers/has-capability.ts

```diff
@@ -9,46 +9,62 @@
 import { capitalize } from '@ember/string';
 
 import type { Capabilities, CapabilitiesMap, CapabilityTypes } from 'vault/app-types';
+import type CapabilitiesService from 'vault/services/capabilities';
+import type { PATH_MAP } from 'vault/utils/constants/capabilities';
 
-type NamedArgs = {
-  id: string;
-  all?: boolean;
-};
-
-export function hasCapability(
-  capabilitiesMap?: CapabilitiesMap,
-  types?: CapabilityTypes[],
-  id?: string,
-  all?: boolean
-) {
-  if (capabilitiesMap && Array.isArray(types)) {
-    const path = Object.keys(capabilitiesMap).find((key) => (id ? key.includes(id) : false));
-
-    if (path) {
-      const capabilities = capabilitiesMap[path];
-
-      if (capabilities) {
-        const method = all ? 'every' : 'some';
-
-        return types[method]((type) => {
-          const key = `can${capitalize(type)}` as keyof Capabilities;
-          // default to true if type provided is not valid - edit rather than update for example
-          return !(key in capabilities) ? true : capabilities[key];
-        });
+export default class HasCapabilityHelper extends Helper {
+  @service declare readonly capabilities: CapabilitiesService;
+
+  compute<T>(
+    [capabilitiesMap, ...types]: [CapabilitiesMap, ...CapabilityTypes[]],
+    { pathKey, params, all = false }: { pathKey: keyof typeof PATH_MAP; params?: T; all?: boolean }
+  ) {
+    return () => {
+      // since the default is to return true we need to validate the inputs here more thoroughly
+      // this is to help devs from thinking that the checks are working properly when the capability lookup is actually failing
+      if (!capabilitiesMap) {
+        throw new Error('Capabilities map is required as the first positional arg');
+      }
+      if (!types || !types.length) {
+        throw new Error('Capability types are required as positional args');
+      }
+      const acceptedTypes = ['read', 'update', 'delete', 'list', 'create', 'patch', 'sudo'];
+      const invalidTypes = types.filter((type) => !acceptedTypes.includes(type));
+      if (invalidTypes.length) {
+        throw new Error(
+          `Invalid capability types: ${invalidTypes.join(', ')}. Accepted types are: ${acceptedTypes.join(
+            ', '
+          )}.`
+        );
       }
-    }
+      if (!pathKey) {
+        throw new Error('pathKey is a required named arg for path lookup in capabilities map');
+      }
+
+      const path = this.capabilities.pathFor(pathKey, params);
+      if (path) {
+        const capabilities = capabilitiesMap[path];
+
+        if (capabilities) {
+          const method = all ? 'every' : 'some';
+
+          return types[method]((type) => {
+            const key = `can${capitalize(type)}` as keyof Capabilities;
+            // default to true if type provided is not valid - edit rather than update for example
+            return !(key in capabilities) ? true : capabilities[key];
+          });
+        }
+      }
```

Comment:
```suggestion
  const capabilities = path ? capabilitiesMap[path] : null;

    if (capabilities) {
      const method = all ? 'every' : 'some';
      return types[method]((type) => {
        const key = `can${capitalize(type)}` as keyof Capabilities;
        // If the capability type is not defined, default to default to true 
        // e.g., if a dev uses 'edit' instead of the valid 'update', this will default to true;
        return key in capabilities ? capabilities[key] : true;
      });
    }
```

Reply:
This suggestion made me realize that some of this is outdated since I added in the extra error checks. Now it will throw an error if the type is not supported (edit instead of update for example). I took the suggestion and removed the `key in capabilities` check.